### PR TITLE
Validate source mappings and option values for lists / typeaheads

### PR DESCRIPTION
**Pull request by pedroslopez**
## Change description

Provides better validation around mappings and list/typeahead options. This is specially useful for validating that columns set on options and mappings actually exist in the source.

Previously, columns for table source properties required at least one record in the system to validate by attempting an import. There was no external validation for source mappings. 

Most of the changes in this PR are within tests to make the new validations not conflict with existing tests.

### Development

- [x] Application changes have been tested appropriately

### Impact

- [x] Code follows company security practices and guidelines
- [x] Security impact of change has been considered
- [x] Performance impact of change has been considered
- [x] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

### CODE REVIEW

File: core/src/modules/mappingHelper.ts [hunk 1]
```diff
@@ -150,4 +154,21 @@ export namespace MappingHelper {
...
+      const validKeys = Object.keys(sourcePreview[0]);
+      if (!validKeys.includes(remoteKey)) {
+        throw new Error(
+          `"${remoteKey}" is not a valid remote mapping key for ${modelName(
+            instance
+          )} ${instance.name}`
+        );
+      }
+    }
+  }
```
Comment (pedroslopez) [inline core/src/modules/mappingHelper.ts:173]: We're using the source preview to know which columns are available to map to for sources. This is equivalent to what the frontend is doing to populate the column selection dropdown.

File: core/src/modules/optionHelper.ts [hunk 2]
```diff
@@ -257,11 +266,21 @@ export namespace OptionHelper {
...
+
+      const val = options[k].toString();
+      const opts = optionOptions[k];
+      if (opts && opts.length > 0 && !opts.includes(val)) {
+        throw new Error(
+          `"${val}" is not a valid value for ${type} ${modelName<ModelWithOptions>(
+            instance
+          )} option "${k}"`
+        );
+      }
```
Comment (pedroslopez) [inline core/src/modules/optionHelper.ts:284]: For options that provide a list of options, we are now checking that the provided value is in that list of options.

**Review by pedroslopez — COMMENTED** [review]
Here are the main changes:

File: core/__tests__/actions/apps.ts [hunk 3]
```diff
@@ -90,12 +90,13 @@ describe("actions/apps", () => {
...
       expect(pluginOptions).toEqual([
         { key: "fileId", required: true },
+        { key: "environment", required: false },
         { key: "password", required: false },
         { key: "oAuthToken", type: "oauth-token", required: false },
         { key: "_failRemoteValidation", required: false },
       ]);
       expect(options).toEqual({
-        fileId: { options: ["a", "b"], type: "list" },
+        environment: { options: ["staging", "production"], type: "list" },
```
Comment (evantahler) [inline core/__tests__/actions/apps.ts:99]: This change captures the problem in a nutshell!

File: plugins/@grouparoo/spec-helper/src/lib/specHelper.ts [hunk 4]
```diff
@@ -410,24 +423,25 @@ export namespace helper {
...
                 description: "the column to choose",
                 type: "list",
-                options: async () => {
-                  return [
-                    { key: "id", examples: [1, 2, 3] },
-                    { key: "fname", examples: ["mario", "luigi", "peach"] },
-                    { key: "lname", examples: ["mario", "mario", "toadstool"] },
-                  ];
+                options: async ({ sourceOptions }) => {
+                  const rows = TestPluginData[sourceOptions.tableWithOptions];
```
Comment (evantahler) [inline plugins/@grouparoo/spec-helper/src/lib/specHelper.ts:436]: 👍 nice

File: core/src/models/Schedule.ts [hunk 5]
```diff
@@ -151,8 +151,8 @@ export class Schedule extends CommonModel<Schedule> {
     return OptionHelper.getOptions(this, sourceFromEnvironment);
   }
 
-  async setOptions(options: SimpleScheduleOptions) {
-    return OptionHelper.setOptions(this, options);
+  async setOptions(options: SimpleScheduleOptions, externallyValidate = true) {
+    return OptionHelper.setOptions(this, options, externallyValidate);
```
Comment (evantahler) [inline core/src/models/Schedule.ts:155]: lots more `externallyValidate(bool)` which makes sense

File: plugins/@grouparoo/mailchimp/__tests__/integration/mailchimpImport.ts [hunk 6]
```diff
@@ -417,9 +417,9 @@ describe("integration/runs/mailchimp-import", () => {
 
         await run.updateTotals();
         expect(run.state).toBe("complete");
-        expect(run.importsCreated).toBe(10);
+        expect(run.importsCreated).toBe(20);
         expect(run.recordsCreated).toBe(0);
-        expect(run.recordsImported).toBe(5);
+        expect(run.recordsImported).toBe(10);
```
Comment (evantahler) [inline plugins/@grouparoo/mailchimp/__tests__/integration/mailchimpImport.ts:422]: The mailchimp importer was always sort of wonky, and dependent on the test data that we keep seeming to delete 😢

File: core/src/modules/optionHelper.ts [hunk 7]
```diff
@@ -18,15 +18,15 @@ import { SourcesCache } from "../modules/caches/sourcesCache";
 
 export const ObfuscatedOptionString = "__ObfuscatedOption";
 
-type ModelWithMapping = Source | Destination | Schedule | Property | App;
+type ModelWithOptions = Source | Destination | Schedule | Property | App;
```
Comment (evantahler) [inline core/src/modules/optionHelper.ts:21]: Oops!

File: core/src/modules/optionHelper.ts [hunk 8]
```diff
@@ -323,6 +342,57 @@ export namespace OptionHelper {
...
+    } else if (instance instanceof App) {
+      const appOptions = await instance.appOptions();
+      return optionOptionsFromObject(appOptions);
+    } else if (instance instanceof Source) {
+      const sourceConnectionOptions = await instance.sourceConnectionOptions(
+        options
+      );
+      return optionOptionsFromObject(sourceConnectionOptions);
+    }
+  }
```
Comment (evantahler) [inline core/src/modules/optionHelper.ts]: Now we've got 2 big case statements in this helper (this + validateOptions) that differ per model.  This sort of looks like we should be doing less in this helper and more per-model maybe?

**Review by evantahler — CHANGES_REQUESTED** [review]
Lots of heroic test updates here! 

I think we should think about the optionHelper a little critically here - it might be doing too much work to support multiple models

File: plugins/@grouparoo/mailchimp/__tests__/integration/mailchimpImport.ts (hunk 6 above)
Comment (pedroslopez) [reply plugins/@grouparoo/mailchimp/__tests__/integration/mailchimpImport.ts:422]: yeah... i was really confused when there was nothing there- same happened with snowflake, too 🤔 
There were also two missing custom fields in mailchimp 🤷🏼 

But I _think_ these should be the values for what the `.env.example` states and the sample data provided in  `data/profiles.csv`

File: core/src/modules/optionHelper.ts (hunk 8 above)
Comment (pedroslopez) [reply core/src/modules/optionHelper.ts]: Yeah, that's a good point. These are _similar_ but they each have slight differences.

I refactored the `validateOptions` in dfb3edf / 5b82795 to be called from the model so we can get the data in the right format and then pass it to the common validation logic.

**Review by evantahler — APPROVED** [review]
Yeah, moving the `validateOptions` methods into the models is less convoluted!

Approved as-is, but a question:  Is it /always/ the case that if a validateOptions method exists, it should always be called as the first step in `OptionHelper.setOptions?`. If so, it might be nice to have that done automatically (vs in each models setOptions) method

**Pull request merged**